examples/error_recovery.rs

```diff
@@ -0,0 +1,67 @@
+// This example illustrates the error flow of a Request in BeforeMiddleware.
+// Here is the chain used and the path of the request through the middleware pieces:
+//
+// Normal Flow : ____[ErrorProducer::before]_____     [ErrorRecover::before]     ____[HelloWorldHandler::handle]___
+// Error Flow  :     [ErrorProducer::catch ]     |____[ErrorRecover::catch ]_____|
+
+
+extern crate iron;
+
+use iron::prelude::*;
+use iron::status;
+use iron::{Handler, BeforeMiddleware};
+
+use std::error::Error;
+use std::fmt::{self, Debug};
+
+struct HelloWorldHandler;
+struct ErrorProducer;
+struct ErrorRecover;
+
+#[derive(Debug)]
+struct StringError(String);
+
+impl fmt::Display for StringError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
+impl Error for StringError {
+    fn description(&self) -> &str { &*self.0 }
+}
+
+impl Handler for HelloWorldHandler {
+    fn handle(&self, _: &mut Request) -> IronResult<Response> {
+        Ok(Response::with((status::Ok, "Hello world !")))
+    }
+}
+
+impl BeforeMiddleware for ErrorProducer {
+    fn before(&self, _: &mut Request) -> IronResult<()> {
+        // The error produced here switches to the error flow.
+        // The catch method of following middleware pieces will be called.
+        // The Handler will be skipped unless the error is handled by another middleware piece.
+        Err(IronError::new(StringError("Error in ErrorProducer".to_string()), status::BadRequest))
```

Comment:
I would also mention here that the `status::BadRequest` is acting as a modifier and that `IronError` contains a `Response` which will be sent if the error is not otherwise handled.